### ISSUE TITLE
Constants slower than literals in range creation

### ISSUE THREAD

**Issue by Michael-S** [rakudo/rakudo#2138]
## The Problem
When using constants in range creation and mathematical operations, the execution seems to be much slower than when using literals.  

## Expected Behavior
I don't know the Rakudo/Moar internals, but I would expect the overhead for dealing with constants vs. literals to be incredibly low.  I can only really guess at overhead, but I would expect maybe a few percent?

## Actual Behavior
Roughly 4x slower speed.

## Steps to Reproduce
(Edit: I'm battling Markdown and losing.  I'll put my comments on the code below here instead of line with pound/hash signs.)
Below are two simple perl6 scripts. 
This script with literals finishes in 1.9 seconds on my old desktop.

    my $running-total = 0;
    for 0..^10 -> $section {
        my $section-start = $section * 100_000 + 1;
        my $section-end = $section-start + 100_000;
        for $section-start..^$section-end -> $item {
    	        $running-total += $item;
        }
    }
    say $running-total;

This script with constants finishes in 9 seconds on the same desktop.

    constant max-num = 1_000_000;
    constant num-sections = 10;
    constant partition-size = max-num / num-sections;
    my $running-total = 0;
    for 0..^num-sections -> $section {
        my $section-start = $section * partition-size + 1;
        my $section-end = $section-start + partition-size;
        for $section-start..^$section-end -> $item {
        	    $running-total += $item;
        }
    }
    say $running-total; 

Timing done with bash/time and GNU time.

## Environment
* Operating system:  Ubuntu MATE 18.04 64-bit, 16GB of RAM, AMD quad-core CPU.
* Compiler version (`perl6 -v`): 
"This is Rakudo version 2018.06 built on MoarVM version 2018.06 implementing Perl 6.c."  Similar but slightly slower performance for both sets of code experienced in Rakudo 2018.04 Perl 6.c.

**Comment by lizmat** [comment]
This is probably because `for ^n { }` is highly optimized.  And I guess the optimization doesn't fire if `n` happens to be a defined constant.  Looking into it.

**Comment by lizmat** [comment]
Yeah, confirmed.  The `constant` case actually creates the `Range` and runs the iterator on it.  The literal case doesn't create the `Range` at all, but just rewrites the loop in `nqp::` ops.

**Comment by Michael-S** [comment]
Wow.  Now that's a fast response.  Thanks.

I don't know how often someone would have headaches with this in the "real world", I was just writing toy programs in Perl6 and stumbled across it.  But I thought it would be worth opening an issue.

**Comment by zoffixznet** [comment]
> the optimization doesn't fire if n happens to be a defined constant

Can we optimize the case of constants as well?

Looking at the QAST, the constant case gets replaced with `WVal (Int)`, but without the `Ii` alternative that https://github.com/rakudo/rakudo/blob/08b449e1ab36ad18c781ed2f67a3fbde59bebb2e/src/Perl6/Optimizer.nqp#L1026-L1033 is looking for. 

Can we just gut that `Int` for its value if `Ii` is missing or is there no way to check whether that `Int` is a core `Int` without any mixins in it?

**Comment by lizmat** [comment]
I think we can optimize the case of constants as well, and should  :-)

**Comment by lizmat** [comment]
Fixed for the case `my constant \foo = 42` by c603340120 .  The case of `my constant $foo = 42` needs to be handled differently.

**Comment by zoffixznet** [comment]
> my constant $foo = 42 needs to be handled differently.

There's a related ticket for that: https://github.com/rakudo/rakudo/issues/1829

**Comment by lizmat** [comment]
Yeah, that looks like the one...  Maybe we can get two tickets resolved for the price of one!

**Comment by lizmat** [comment]
Ok, I'm giving up on this one.

For some reason, when you define a constant with a sigil `my constant $foo = 42`, it winds up not being a constant in the `for ^$foo { }` case: instead of a `QAST::Want` for a literal, or a `QAST::WVal` for a sigilless constant, it gets a `QAST::Var`.  FWIW, I think that's wrong.  If something is a constant, it should be codegenned as a `QAST::WVAL`, regardless of the sigil.

Looking at the `%sym` info of the (variable) constant, there does *not* seem to be a way to distinguish this from a *real* variable.

If I look at the `method type_declarator:sym<constant>($/)` in Actions, I don't detect anything that would handle a sigilless constant from a constant with a sigil.  The only thing I *do* see that if a constant is defined with a sigil, it *is* run through the `check_variable` method in `Grammar`.

I've now spent the better part of a day on this, and have not come any further.  For now, I would suggest if you want to use constants in `for` loops, make sure they **don't** have a sigil.

Also unassigning myself.

**Comment by Michael-S** [comment]
Thanks for tackling it.  I'll remember that, constants shouldn't use sigils.

I leave it up to the project maintainers whether they close the ticket or leave it open to revisit the issues lizmat brought up in the previous comment.
